### PR TITLE
feat: Add backup and restore functionality to k8s

### PR DESCRIPTION
**Pull request by foadlind**
* Add Tutor configuration parameters to specify a S3 bucket for
  backup storage.
* Add `backup` and `restore` sub-commands to k8s command group. The
  backup commands runs the backup job that uploads the tarball to S3,
  which the restore job downloads when invoked.
* By default the backup job runs according to the schedule provided
  by `BACKUP_K8S_CRONJOB_BACKUP_SCHEDULE`. If this is set to `None`,
  no CronJob is created.
* Unless the users sets `BACKUP_K8S_CRONJOB_RESTORE_SCHEDULE`, no
  CronJob is added for restore.
* If a version ID is supplied to the restore command using the
  `--version` argument, the restore job downloads that version
  of the backup, otherwise it downloads the latest version.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,12 +47,88 @@ a dump of the MongoDB database, and a copy of the Caddy data directory.
...
+the scheduled job completely, set `BACKUP_K8S_CRONJOB_BACKUP_SCHEDULE` to None. 
+Note that you need to restart your Kubernetes deployment with 
+`tutor k8s quickstart` for this change to take effect.
+
+If you need to run the backup job outside the schedule, use:
+
+    tutor k8s backup
+
+The backup job backs up MySQL, MongoDB, and the Caddy data directory, 
+creates a tar fil, and uploads it to an S3 storage bucket as set by
```
Comment (fghaas) [inline README.md:72]: Micro-nit: missing "e" in "file". :)

File: tutorbackup/patches/k8s-jobs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+        - labelSelector:
+            matchExpressions:
+            - key: app.kubernetes.io/name
+              operator: In
+              values:
+                - caddy
+          topologyKey: topology.kubernetes.io/zone
```
Comment (fghaas) [inline tutorbackup/patches/k8s-jobs:81]: Are we sure we don't want `topologyKey: "kubernetes.io/hostname"` here?

**Review by fghaas — COMMENTED** [review]
This looks good overall! I have a few questions/suggestions.

File: README.md (hunk 1 above)
Comment (foadlind) [reply README.md:72]: will fix

File: tutorbackup/plugin.py [hunk 3]
```diff
@@ -15,6 +16,19 @@
     "defaults": {
         "VERSION": __version__,
         "DOCKER_IMAGE": "{{ DOCKER_REGISTRY }}backup:{{ BACKUP_VERSION }}",  # noqa: E501
+        "K8S_CRONJOB_HISTORYLIMIT_FAILURE": 1,
+        "K8S_CRONJOB_HISTORYLIMIT_SUCCESS": 3,
+        "K8S_CRONJOB_BACKUP_SCHEDULE": "0 0 * * *",
+        "K8S_CRONJOB_RESTORE_SCHEDULE": None,
+        "S3_HOST": None,
```
[the remark this answers is not in the document]
Comment (foadlind) [reply tutorbackup/plugin.py:23]: That's a good idea. I'll do that

File: tutorbackup/patches/k8s-jobs (hunk 2 above)
Comment (fghaas) [reply tutorbackup/patches/k8s-jobs:81]: The `topology.kubernetes.io` label maps to an availability zone. `kubernetes.io/hostname` maps to the host name, which I believe is what we want here.

**Pull request merged**